docs/plugin-write.md

```diff
@@ -0,0 +1,242 @@
+type Plugin struct {
+    config *Config
+}
+
+type Config struct {
+    GitHubUserName string
+}
+
+// DefaultConfig implements plugin.Configurer
+// The default configuration will be provided to the user for future editing. Also used for Unmarshaling.
+// Invoked whenever an unmarshaling is required.
+func (c *Plugin) DefaultConfig() interface{} {
+    return &Config{
+        GitHubUserName: "jmattheis",
+    }
+}
+
+// ValidateAndSetConfig will be called every time the plugin is initialized or the configuration has been changed by the user.
+// Plugins should validate the configuration and optionally return an error.
+// Parameter is guaranteed to be the same type as the return type of DefaultConfig(), so it is safe to do a hard type assertion here.
+//
+// "Validation" in this context means to check for conflicting or impossible values, such as a non-URL on a field which should only contain a URL.
+// In order to make sure that the plugin instance is always running in a valid state, this method should always accept the result of DefaultConfig()
+//
+// Invoked on initialization to provide initial configuration. Return nil to accept or return error to indicate that the config is obsolete.
+// When the configuration is marked obsolete due to an unmarshaling error or rejection on the plugin side, the plugin is disabled automatically and the user is notified to resolve the config confliction.
+// Invoked every time the config update API is called. Check the configuration and return nil to accept or return error to indicate that the config is invalid.
+// Return a short and consise error here and, if you have detailed suggestions on how to solve the problem, utilize Displayer to provide more information to the user,
+func (c *Plugin) ValidateAndSetConfig(c interface{}) error {
+    config = c.(*Config)
+    if !userNameIsValid(config.GitHubUserName) {
+        return errors.New("the user name is not valid")
+    }
+    c.config = config
+    return nil
+}
+```
+
+## Effective Plugin Practices
+
+Although we have covered how to implement plugin functionalities in the last chapter, it is very important to take these practices to make sure your plugin can be loaded successfully and works effectively.
+
+- Use [go modules](https://github.com/golang/go/wiki/Modules) to manage dependencies, and use [gomod-cap](https://github.com/gotify/plugin-api/#githubcomgotifycmdgomod-cap) to prevent incompatible dependencies.
+- Handle all errors. A panic in a goroutine spawned in the plugin can crash the whole gotify program.
+- Provide detailed plugin info and utilize [`Displayer`](/docs/plugin-api#displayer) to show instructions to users. A detailed plugin info would be shown in the WebUI which makes it easier to be identified and used.
+
+You can clone the official plugin [template](https://github.com/gotify/plugin-template) and [showcase](https://github.com/gotify/plugins) to see plugins in action and/or to bootstrap your project.
```

Comment:
gotify/plugins will be a metadata repository for existing plugins, a searchable list will be published to https://plugins.gotify.net/. It should be like https://galaxy.ansible.com/.

Just FYI (:.

Reply:
Okay, but I would still use the GitHub link as this is more developer oriented so maybe a GitHub like directly to the source code would be more helpful:)